specs/phase0/weak-subjectivity.md

```diff
@@ -59,38 +77,81 @@ a safety margin of at least `1/3 - SAFETY_DECAY/100`.
+def compute_avg_active_validator_balance(state: BeaconState) -> Ether:
+    total_active_balance = get_total_active_balance(state)
+    active_validator_count = get_active_validator_count(state)
+    avg_active_validator_balance_gwei = total_active_balance // active_validator_count
+    avg_active_validator_balance_eth = avg_active_validator_balance_gwei // ETH_TO_GWEI
+    return avg_active_validator_balance_eth
+```
+
+#### `compute_weak_subjectivity_period`
 
 ```python
 def compute_weak_subjectivity_period(state: BeaconState) -> uint64:
-    weak_subjectivity_period = MIN_VALIDATOR_WITHDRAWABILITY_DELAY
-    validator_count = len(get_active_validator_indices(state, get_current_epoch(state)))
-    if validator_count >= MIN_PER_EPOCH_CHURN_LIMIT * CHURN_LIMIT_QUOTIENT:
-        weak_subjectivity_period += SAFETY_DECAY * CHURN_LIMIT_QUOTIENT // (2 * 100)
+    ws_period = MIN_VALIDATOR_WITHDRAWABILITY_DELAY
+    N = get_active_validator_count(state)
+    t = compute_avg_active_validator_balance(state)
+    T = MAX_EFFECTIVE_BALANCE // 10**9
+    delta = get_validator_churn_limit(state)
+    Delta = MAX_DEPOSITS * SLOTS_PER_EPOCH
+    D = SAFETY_DECAY
+
+    case = (
+        T * (200 + 3 * D) < t * (200 + 12 * D)
+    )
+
+    if case == 1:
+        arg1 = (
+            N * (t * (200 + 12 * D) - T * (200 + 3 * D)) // (600 * delta * (2 * t + T))
+        )
+        arg2 = (
+            N * (200 + 3 * D) // (600 * Delta)
+        )
+        ws_period += max(arg1, arg2)
     else:
-        weak_subjectivity_period += SAFETY_DECAY * validator_count // (2 * 100 * MIN_PER_EPOCH_CHURN_LIMIT)
-    return weak_subjectivity_period
+        ws_period += (
+            3 * N * D * t // (200 * Delta * (T - t))
+        )
+    
+    return ws_period
 ```
 
-*Details about the calculation*:
-- `100` appears in the denominator to get the actual percentage ratio from `SAFETY_DECAY`
-- For more information about other terms in this equation, refer to
-  [Weak Subjectivity in Eth2.0](https://notes.ethereum.org/@adiasg/weak-subjectvity-eth2)
-
 A brief reference for what these values look like in practice:
 
-| `validator_count` | `weak_subjectivity_period` |
-| ----  | ---- |
-| 1024  | 268 |
-| 2048  | 281 |
-| 4096  | 307 |
-| 8192  | 358 |
-| 16384 | 460 |
-| 32768 | 665 |
-| 65536 | 1075 |
-| 131072  | 1894 |
-| 262144  | 3532 |
-| 524288  | 3532 |
+| Safety Decay | Validator Count | Average Active Validator Balance | Weak Subjectivity Period |
+| ---- | ---- | ---- | ---- |
+| 10 | 8192 | 28 | 318 |
+| 10 | 8192 | 32 | 358 |
+| 10 | 16384 | 28 | 380 |
+| 10 | 16384 | 32 | 460 |
+| 10 | 32768 | 28 | 504 |
+| 10 | 32768 | 32 | 665 |
+| 20 | 8192 | 28 | 411 |
+| 20 | 8192 | 32 | 460 |
+| 20 | 16384 | 28 | 566 |
+| 20 | 16384 | 32 | 665 |
+| 20 | 32768 | 28 | 876 |
+| 20 | 32768 | 32 | 1075 |
+| 33 | 8192 | 28 | 532 |
+| 33 | 8192 | 32 | 593 |
+| 33 | 16384 | 28 | 808 |
+| 33 | 16384 | 32 | 931 |
+| 33 | 32768 | 28 | 1360 |
+| 33 | 32768 | 32 | 1607 |
```

Comment:
I recommend less granularity on the low end and adding some much larger numbers. We're already at 70k active vals and less 16k doesn't seem likely *ever*